docs/PROTOCOL_v21.md

```diff
@@ -0,0 +1,375 @@
+
+### Result Store
+
+This component is responsible for storing upkeepResults that a node thinks should be performed. It hopes to get consensus from the network on these results to push into reports. Best effort is made to ensure the same logs enter different node’s result stores independently as **it is assumed that blockchain nodes will get in sync within TTL**, but it is not guaranteed as node’s local blockchain can see different reorgs or select different logs during surges. For results that do not achieve consensus within TTL go into recovery flow.
+
+- Maintains an in-memory collection of upkeepResults. It should maintain a single result per `upkeepTriggerID`. 
+    - Overwrites results for duplicated `upkeepTriggerID`.
+- Each result has a TTL. When the TTL expires
+    - conditional upkeeps will be sampled and checked again automatically
+    - log upkeeps that were missed are expected to be picked up by the log recoverer **TBD**
+- Provides a read API to view results, which takes as input (`pseudoRandomSeed`, `limit`). Sorts all keys (trigger, upkeepID) with the `seed` and provides first `limit` results. We do not do FIFO here as potentially out of sync old results will block new results sent by the node for consensus.
+- Provides an API to remove (trigger, upkeepID) from the store
+
+### Metadata Store
+
+This component stores the metadata to be sent within observations to the network. There are three categories of metadata
+
+- Conditional Upkeep Instructions
+- Log recovery Instructions
+- Latest block history
+
+Every item has a TTL, and upon expiry items are just removed without any action.
+
+The store provides an add / view / remove API for other components.
+
+### Samples Observer
+
+Processes samples of upkeeps to be checked, using the latest block number as a trigger. It does the following procedures:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **metadata store**
+with conditional Upkeep instructions
+- Errors and ineligible results are ignored
+
+### Conditional Observer
+
+Processes coordinated block + upkeep payload coming from plugin and does the following:
+
+- Pre-processes to filter upkeep present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into **result store**
+- Errors and ineligible results are ignored
+
+### Log Observer
+
+Processes latest logs for active upkeeps. It does the following procedures:
+
+- Called every second by Log Trigger Ticker with latest logs as full payloads.
+- Filter out logs with block number that is older than threshold
+    - These are expected to be picked up by the log recoverer 
+- Pre-processes to filter the logs already present in coordinator
+- Calls runner with upkeep payload
+- If upkeep is eligible enqueue into result store
+- If runner gave an error, put back into log retry ticker, to be performed after a fixed timeout (can extend to exponential backoff)
```

Comment:
Exponential backoff isn't necessary here as the retry window is very small anyway. The retry cadence is to wait 10 seconds, try again, and repeat for up to 5 minutes. This is configurable.

Reply:
How does this work when block goes out of range? (e.g. on arbitrum we can only retry for 128 blocks ~60 seconds)